src/DataAPI.jl

```diff
@@ -287,4 +287,35 @@ using a `sink` function to materialize the table.
 """
 function allcombinations end
 
+"""
+    metadata(x, [key])
+
+Return metadata associated with object `x` as an `AbstractDict{String}` object
+(or an object implementing the same interface).
+
+If the optional `key` argument is passed, return metadata
+associated with the indicated key (this feature can be used, for example,
+to return metadata attached to a given column in a Tables.jl table).
+
+Note that some systems, like Arrow.jl, might assume that metadata values are
+also `String`.
+"""
+metadata(::T) where {T} =
+    throw(ArgumentError("Metadata is currently not supported for values of type $T"))
+metadata(::T, ::Any) where {T} =
+    throw(ArgumentError("Key metadata is currently not supported for values of type $T"))
```

Comment:
This fallback can give misleading errors if you pass the wrong argument. For example:
```julia
julia> metadata(DataFrame(x=1), 1.0)
ERROR: ArgumentError: Key metadata is currently not supported for values of type DataFrame
```

Maybe something like this would be clearer?
```suggestion
metadata(::T) where {T} =
    throw(ArgumentError("Objects of type $T do not support metadata"))
metadata(::T, ::U) where {T, U} =
    throw(ArgumentError("Objects of type $T do not support metadata for keys of type $U"))
```

Reply:
Would we be assuming too much if we just did `metadata(x, key) = metadata(x)[key]`? Then we'd fall back to the appropriate error if there's no metadata and get the appropriate error of indexing by key isn't supported.

Reply:
`metadata(x, key)` is a metadata dictionary associated with `key`,
while
`meatadata(x)[key]` is a value associated with `key` key for metadata dictionary associated with `x`.

In short `metadata(x, key)` and `metadata(x)` are separate metadata dictionaries (here `key` serves as a kind-of additional metadata scope specifier).

Reply:
Oh, so you would use this to do something like `metadata(x, :column)`?

Reply:
Yes - in Tables.jl table context. However, I assume that in different contexts `key` can mean something different.

Reply:
I think `metadata(frame, column)` API would be a constant source of confusion with people assuming it is `metadata(frame, key)`.
What about `colmetadata(frame, column)` or allowing `metadata(frame.column)`?
The latter, of course, would require that `frame.column` returns some thin wrapper around column data, which is quite a change.

Reply:
> `frame.column` returns some thin wrapper around column data

The whole point of this API extension is that this thin wrapper is very problematic and would confuse users in other contexts.

`colmetadata` is OK I think, then also `colmetadata(frame)` could return a collection of all column metadata defined for the whole table. @nalimilan - what do you think?

Reply:
Good point. I was also concerned that `metadata(df, col)` isn't consistent with `getindex(df, :, col)`. Using `colmetadata` is more explicit and it avoids the possibility of conflicts between different interfaces.

Reply:
Since it came up, my recent comment [here](https://github.com/JuliaData/DataAPI.jl/issues/22#issuecomment-1137755587) may be of interest. When the context we assign to metadata (what it is attached to) is an in memory object there's a reasonably small grouping we can create for referring to metadata and its relationship to data. I'm not sure how much more you care to add to this PR or to subsequent PRs.

Reply:
It seems like we'd better start with the basic API proposed by this PR, and later extend it to cover more complex needs. IIUC in your proposal `metadata(x)` corresponds to self metadata, and `colmetadata(x, col)` to axis metadata (columns are the second axis).

Reply:
That's essentially correct but I haven't done any bike-shedding with regards to the exact syntax, just the conceptual hierarchy.